Implement stable version of `cudf::sort`

wence- opened the pull request:
## Description

Adds an implementation of `cudf::stable_sort`. While here, cleans up a few small issues around stream-passing and memory resource usage in the detail APIs of some of the sort functions.

- Closes #15065

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

wence- commented:
I suppose I should add some tests. I can write ones that check that the result is correct, but is there any way to check that the non-stable sort would not have produced that result?

wence- commented:
A single round of the cudf-python tests failed due to not raising pandas warnings where we expect to. But I would have thought this should also happen on different CTK versions, so no idea what is going on.

davidwendt commented:
> I suppose I should add some tests. I can write ones that check that the result is correct, but is there any way to check that the non-stable sort would not have produced that result?

Although you cannot really verify the results are _stable_, tests are helpful in case this diverges in the future and we want to make sure it doesn't just fail sorting in general.

wence- commented on cpp/src/sort/stable_sort.cu:
```diff
@@ -37,6 +37,28 @@ std::unique_ptr<column> stable_sorted_order(table_view const& input,
+                                   rmm::mr::device_memory_resource* mr)
+{
+  // fast-path sort conditions: single, non-floating-point, fixed-width column with no nulls
+  if (input.num_columns() == 1 && !input.column(0).has_nulls() &&
```
Done in `"common_sort_impl.cuh"`. I also took the opportunity to unify all templated sort calls as taking a `sort_method` enum template parameter, rather than a `bool` to indicate stable sort.

wence- commented on cpp/src/sort/sort.cu:
```diff
@@ -61,47 +62,24 @@ std::unique_ptr<table> sort_by_key(table_view const& values,
-    auto output    = std::make_unique<column>(input.column(0), stream, mr);
-    auto view      = output->mutable_view();
-    bool ascending = (column_order.empty() ? true : column_order.front() == order::ASCENDING);
+  if (can_use_fast_sort(input)) {
```
Perhaps rename as `can_use_inplace_sort` ?

wence- commented on cpp/tests/sort/stable_sort_tests.cpp:
```diff
@@ -34,12 +34,14 @@
+                          std::vector<cudf::null_order> null_precedence = {},
+                          bool by_key                                   = true)
 {
-  auto got_sort_by_key_table      = cudf::sort_by_key(input, input, column_order, null_precedence);
```
Many of the stable sort tests were previously not testing a stable sort!

wence- commented on cpp/tests/sort/stable_sort_tests.cpp:
```diff
@@ -78,34 +80,58 @@ TYPED_TEST(StableSort, WithNullMax)
-
-    run_stable_sort_test(input, expected, column_order, null_precedence);
-  } else {
-    // for bools only validate that the null element landed at the back, since
-    // the rest of the values are equivalent and yields random sorted order.
```
This comment was only true for the unstable sort, which these tests are not.

wence- commented on cpp/tests/sort/stable_sort_tests.cpp:
```diff
@@ -78,34 +80,58 @@ TYPED_TEST(StableSort, WithNullMax)
-  }
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected, got->view());
+
+  run_stable_sort_test(input, expected, column_order, null_precedence, false);
+  run_stable_sort_test(input, expected, column_order, null_precedence, true);
```
I don't know how to value-parametrize a typed test to run them separately with `by_key = true` and `by_key = false`, so just running both versions here.

wence- commented on cpp/tests/sort/stable_sort_tests.cpp:
```diff
@@ -78,34 +80,58 @@ TYPED_TEST(StableSort, WithNullMax)
+  run_stable_sort_test(input, expected, column_order, null_precedence, true);
+}
+
+TYPED_TEST(StableSort, SingleColumnNoNull)
```
This test exercises the "fast-path" single column sort.

davidwendt replied:
> This test exercises the "fast-path" single column sort.

Can you add this comment to the test source code here?

wence- replied:
Done.

wence- commented on cpp/src/sort/segmented_sort_impl.cuh:
```diff
@@ -29,11 +31,6 @@
 namespace cudf {
 namespace detail {
 
-/**
- * @brief The enum specifying which sorting method to use (stable or unstable).
- */
-enum class sort_method { STABLE, UNSTABLE };
-
```
Moved to common_sort_impl.cuh

wence- commented on cpp/src/sort/segmented_sort_impl.cuh:
```diff
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
```
```suggestion
#pragma once

```

wence- commented on cpp/src/sort/sort.cu:
```diff
@@ -37,7 +38,7 @@ std::unique_ptr<column> sorted_order(table_view const& input,
                                      rmm::mr::device_memory_resource* mr)
 {
-  return sorted_order<false>(input, column_order, null_precedence, stream, mr);
+  return sorted_order<sort_method::UNSTABLE>(input, column_order, null_precedence, stream, mr);
```
While I was in the area, moving to consistently using the enum-based approach, since it is slightly more self-documenting at the call site.

wence- commented:
> Although you cannot really verify the results are _stable_, tests are helpful in case this diverges in the future and we want to make sure it doesn't just fail sorting in general.

Added tests, and fixed a two-year-old bug that meant that we weren't testing stable_sort_by_key in a bunch of the stable_sort tests.

davidwendt reviewed (commented):
This looks great. Thanks for fixing the tests too.

wence- commented on cpp/src/sort/common_sort_impl.cuh:
```diff
@@ -0,0 +1,96 @@
+template <sort_method method>
+struct inplace_column_sort_fn {
+  /**
+   * @brief Can one use a fast-path in-place sort for this column?
```
Thanks, done (and for the table-dispatch version too)

davidwendt commented on cpp/src/sort/common_sort_impl.cuh:
```diff
@@ -0,0 +1,96 @@
+ * Precondition, is_usable(column) returned true
+ *
+ * @tparam method Whether to use a stable sort or not.
+ * @param col Column to sort, modified in place.
+ * @param order Ascending or descending sort order.
+ * @param stream CUDA stream used for device memory operations and kernel launches
```
These should be with the `operator()` function instead of the class doxygen.

wence- replied:
Done, thanks.

davidwendt commented on cpp/src/sort/common_sort_impl.cuh:
```diff
@@ -0,0 +1,96 @@
+template <sort_method method>
+struct inplace_column_sort_fn {
+  /**
+   * @brief Check if fast-path, in place sort is available for the given column.
```
```suggestion
   * @brief Check if fast-path, in-place sort is available for the given column
```

davidwendt reviewed (approved):
I'm good with the names of the enums and the `is_usable()` as well.

wence- commented:
/merge
